transports/quic/src/lib.rs

```diff
@@ -57,16 +57,17 @@
 
 #![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
 
+mod config;
 mod connection;
-mod endpoint;
 mod hole_punching;
 mod provider;
```

Comment:
A user can, by design, not implement `provider::Provider`. With that in mind, do we need to expose the `Provider` trait further below? In other words, would the following diff be valid @thomaseizinger?

``` diff
modified   transports/quic/src/lib.rs
@@ -72,7 +72,6 @@ pub use connection::{Connecting, Connection, Stream};
 pub use provider::async_std;
 #[cfg(feature = "tokio")]
 pub use provider::tokio;
-pub use provider::Provider;
 pub use transport::GenTransport;
 
 /// Errors that may happen on the [`GenTransport`] or a single [`Connection`].
@@ -90,7 +89,7 @@ pub enum Error {
     #[error(transparent)]
     Io(#[from] std::io::Error),
 
-    /// The task spawned in [`Provider::spawn`] to drive
+    /// The task spawned in [`provider::Provider::spawn`] to drive
     /// the quic endpoint has crashed.
     #[error("Endpoint driver crashed")]
     EndpointDriverCrashed,
```

Reply:
We might have to expose it because it appears in a public interface somewhere. I am in favor of hiding it if we can!

Reply:
Did some small experiments. See latest commits. Without success.

Slightly related: I wonder whether `Provider` should rather be an `enum` than a `trait`.  When all potential implementations are known inside the crate and there is no need for extension from outside the crate, I would choose an enum over a trait.

Either way. I am fine leaving it as is or tackling it in a follow-up pull request.